### ISSUE TITLE
[possibly-undefined] False-positive with nested try block and finally

### ISSUE THREAD

**Issue by cdce8p** [python/mypy#14309]
**Bug Report**
Small false-positive I noticed after #14114.
Seems to be a scoping issue with `finally`.

/CC: @ilinum 

**To Reproduce**

```python
try:
    try:
        msg = some_func()
    except Exception:
        raise

    print(msg)  # Name "msg" may be undefined  [possibly-undefined]

finally:
    pass

print(msg)  # <-- should probably be emitted here
```

The first print will always safely execute since the inner Try block raises. If it throws an uncaught exception, the print will be skipped to the `finally` block and only the second print will execute which would then raise an error itself as `msg` would be undefined.

**Your Environment**

- Mypy version used: `mypy 1.0.0+dev.96ac8b3e71c743b02ea4e3c84da0248659e40f82 (compiled: no)`
- Mypy command-line flags: `--enable-error-code possibly-undefined`
- Python version used: 3.11

**Comment by erictraut** [comment]
If `some_func` raises an exception, the second `print` will never be executed because the exception will be re-raised at the end of the `finally` block. So I think that means no error should be generated for the final line of the code sample.

**Comment by ilinum** [comment]
Yep, good find! This only occurs with nested statements because we have to do a ~hack~ clever trick with [disable_branch_skip](https://github.com/python/mypy/blob/d391f108c195533a5800afe73f5b3257babc83a3/mypy/partially_defined.py#L497). 

We can probably do something smarter here and only report errors in `finally` blocks when we're in `disable_branch_skip` mode.

**Comment by Dreamsorcerer** [comment]
Not entirely sure if this is the same issue or a separate issue, but we are also seeing this where an `except` then `break`s from a loop:
https://github.com/aio-libs/aiohttp/blob/94818391872626cd75d354a8afdef0b7bc54d99f/aiohttp/web_protocol.py#L540-L548

**Comment by erictraut** [comment]
Is there still an issue here? If so, could someone provide a self-contained minimal code sample?

**Comment by andersk** [comment]
Here’s the example as a playground link (using [inline configuration](https://mypy.readthedocs.io/en/stable/inline_config.html)):

https://mypy-play.net/?mypy=latest&python=3.11&gist=c74b25457397d60fe5f063151ceb5fee

**Comment by Dreamsorcerer** [comment]
Ah, the presence of the `finally` affects the output. Here's the aiohttp version with a `break`:
```

while True:
    try:
        try:
            msg = some_func()
        except Exception:
            break
        print(msg)
    except Exception:
        pass
    finally:
        pass
```

**Comment by asottile-sentry** [comment]
here's a much simpler example that appears to be an exact duplicate:

```python
def f(x: int) -> None:
    try:
        if x >= 5:
            y = 2
        else:
            raise AssertionError('unreachable')
        print(y)
    finally:
        print('finally!')
```

```console
$ mypy t.py --enable-error-code=possibly-undefined
t.py:7: error: Name "y" may be undefined  [possibly-undefined]
Found 1 error in 1 file (checked 1 source file)
```

**Comment by srittau** [comment]
Another instance from python/typeshed#12151:

```
import sys

try:
    if sys.platform == "win32":
        pip_exe = ""
        reveal_type(pip_exe)  # Revealed type is "builtins.str"
    else:
        pip_exe = ""
    reveal_type(pip_exe)  # Revealed type is "builtins.str"
                                          # Name "pip_exe" may be undefined  [possibly-undefined]
finally:
    pass
```
